src/states/proving_node.rs

```diff
@@ -0,0 +1,555 @@
+    ack_mgr: AckManager,
+    /// ID from before relocating.
+    old_full_id: FullId,
+    full_id: FullId,
+    /// Routing messages addressed to us that we cannot handle until we are approved.
+    msg_backlog: Vec<RoutingMessage>,
+    min_section_size: usize,
+    peer_mgr: PeerManager,
+    cache: Box<Cache>,
+    routing_msg_filter: RoutingMessageFilter,
+    timer: Timer,
+    resource_prover: ResourceProver,
+    /// Whether resource proof is disabled.
+    disable_resource_proof: bool,
+    joining_prefix: Prefix<XorName>,
+    // TODO: notify without local state
+    notified_nodes: BTreeSet<PublicId>,
+}
+
+impl ProvingNode {
+    #[allow(clippy::too_many_arguments)]
+    pub fn from_bootstrapping(
+        our_section: (Prefix<XorName>, BTreeSet<PublicId>),
+        action_sender: RoutingActionSender,
+        cache: Box<Cache>,
+        crust_service: Service,
+        old_full_id: FullId,
+        new_full_id: FullId,
+        min_section_size: usize,
+        proxy_pub_id: PublicId,
+        timer: Timer,
+    ) -> Self {
+        let dev_config = config_handler::get_config().dev.unwrap_or_default();
+        let public_id = *new_full_id.public_id();
+
+        let mut peer_mgr = PeerManager::new(public_id, dev_config.disable_client_rate_limiter);
+        peer_mgr.insert_peer(Peer::new(proxy_pub_id, PeerState::Proxy));
+
+        let challenger_count = our_section.1.len();
+        let resource_prover = ResourceProver::new(action_sender, timer.clone(), challenger_count);
+
+        let mut node = Self {
+            crust_service,
+            ack_mgr: AckManager::new(),
+            old_full_id,
+            full_id: new_full_id,
+            msg_backlog: Vec::new(),
+            min_section_size,
+            peer_mgr,
+            cache,
+            routing_msg_filter: RoutingMessageFilter::new(),
+            timer,
+            resource_prover,
+            disable_resource_proof: dev_config.disable_resource_proof,
+            joining_prefix: our_section.0,
+            notified_nodes: Default::default(),
+        };
+        node.start(our_section.1, &proxy_pub_id);
+        node
+    }
+
+    /// Called immediately after construction. Sends `ConnectionInfoRequest`s to all members of
+    /// `our_section` to then start the candidate approval process.
+    fn start(&mut self, our_section: BTreeSet<PublicId>, proxy_pub_id: &PublicId) {
+        self.resource_prover.start(self.disable_resource_proof);
+
+        trace!("{} Relocation completed.", self);
+        info!(
+            "{} Received relocation section. Establishing connections to {} peers.",
+            self,
+            our_section.len()
+        );
+
+        let src = Authority::Client {
+            client_id: *self.full_id.public_id(),
+            proxy_node_name: *proxy_pub_id.name(),
+        };
+        // There will be no events raised as a result of these calls, so safe to just use a
+        // throwaway `EventBox` here.
```

Comment:
And I guess we can't remove the arguments because they're trait methods?
Could the trait have an associated type instead, that could be set to `EventBox` or `()`?

Reply:
Good idea. I'll give it a try.

Reply:
Unfortunately, this turned out to not be possible. But I added a `outbox` param to the function, so it is now handled normally like anywhere else, removing this weird special case.